File: inst/python/washdev_scraping.py
#!/usr/bin/env python
# coding: utf-8

import bs4
import requests
import os
import datetime
import pickle
import re
import pandas as pd
from tqdm import tqdm


# This script contains the code of scraping the data of Journal of Water, Sanitation and Hygiene for Development. 


# # Overview of all journal issues

# The following section will iterate all issues available online to get information of each issue, namely, 
# the volume and issue number, the accepted paper title and its url link. The journal starts from year 2011 
# as its first volume and made several changes on publication cycles. Since 2022, the journal publishes monthly.

# table of contents 

def get_issue_page(root, volume_id, issue_id, headers = {"User-Agent": "user"}):
    """
    Retrieve an issue html content for WASH Development 
    """
    issue_url = os.path.join(root, str(volume_id), str(issue_id))
    page = requests.get(issue_url, headers=headers)
    status = page.status_code
    if status == 200:
        issue = {
            "volume": volume_id,
            "issue":issue_id,
            "page": page.content
        }
    else:
        print(f"Retrieve vol {volume_id} issue {issue_id} not successful!")
        issue = {
            "volume": volume_id,
            "issue":issue_id,
            "page": ""
        }
    return issue


# Retrieve data of each publication on supplementary materials, author, and data availability statement
# Supplementary data existence
def is_supp(soup):
    """
    True if having supplementary material
    """
    return False if soup.find("h2", id="supplementary-data") is None else True

def get_supp_number(soup):
    """
    Number of supplementary materials
    """
    number_of_supp = len(soup.find_all("div", {"class": "dataSuppLink"}))
    return number_of_supp

def get_supp_file_type(soup):
    """
    File types of the supplementary materials
    """
    supp_list = soup.find_all("div", {"class": "dataSuppLink"})
    if len(supp_list) == 0:
        return None
    elif len(supp_list)==1:
        supp_type = supp_list[0].span.next.next.split(" ")[1]
        return supp_type
    else:
        supp_type_list = []
        for s in supp_list:
            supp_type_list.append(s.span.next.next.split(" ")[1])
        num_type = len(set(supp_type_list))
        if num_type == 1:
            return supp_type_list[0]
        else: # TODO: separate a case for num_type is 2 
            return "misc"
        
def get_supp_file_link(soup):
    """
    URL of supplementary files if available
    """
    supp_list = soup.find_all("div", {"class": "dataSuppLink"})
    link = []
    for s in supp_list:
        link.append(s.a["href"])
    return link
    
def get_supp(soup):
    """
    Retrieve all variables related to supplementary materials
    """
    number_of_supp = get_supp_number(soup)
    supp_file_type = get_supp_file_type(soup)
    supp_file_link = get_supp_file_link(soup)
    supp = [is_supp(soup), number_of_supp, supp_file_type, supp_file_link]
    return supp

# First author (correspondence author) information
def get_author_info(soup):
    """
    Collect metadata of first and correspondence author
    """
    try:
        first_author = soup.find("div", {"class": "info-card-author"})
        if first_author is not None:
            first_author_name = first_author.find("div", {"class": "info-card-name"}).next
            first_author_name = str.strip(first_author_name)
            first_author_has_affiliation = first_author.find("div", {"class": "aff"})
            if first_author_has_affiliation:
                if first_author_has_affiliation.span is not None:
                    first_author_affiliation = first_author_has_affiliation.span.nextSibling.text
                else:
                    first_author_affiliation = first_author_has_affiliation.text
                first_author_affiliation_country = first_author_affiliation.split(", ")[-1]
                first_author_affiliation_country = first_author_affiliation_country.split(" Email")[0]
            else:
                first_author_affiliation, first_author_affiliation_country = None, None
            first_author_has_email = first_author.select('a[href^=mailto]')
            first_author_email = first_author_has_email[0].text if first_author_has_email else None
            first_author_has_orcid = first_author.select('a[id^=contrib-orcid]')
            first_author_orcid = first_author_has_orcid[0]['href'] if first_author_has_orcid else None
            first_author_info = [first_author_name, first_author_affiliation, first_author_affiliation_country, 
                                 first_author_email, first_author_orcid]    
        else:
            return [0, None, None, None, None, None, None, None, None, None, None] # No author information
    except AttributeError:
        print(first_author_has_affiliation.span)
    try:
        corr_author_email_card = soup.find("div", {"class": "info-author-correspondence"})
        if corr_author_email_card is not None:
            corr_author = corr_author_email_card.parent
            corr_author_name = corr_author.find("div", {"class": "info-card-name"}).next
            corr_author_name = str.strip(corr_author_name)
            corr_author_has_aff = corr_author.find("div", {"class": "aff"})
            if corr_author_has_aff:
                if corr_author_has_aff.span is not None:
                    corr_author_affiliation = corr_author_has_aff.span.nextSibling.text
                else:
                    corr_author_affiliation = corr_author_has_aff.text
                corr_author_affiliation_country = corr_author_affiliation.split(", ")[-1]
                corr_author_affiliation_country = corr_author_affiliation_country.split(" Email")[0]
            else:
                corr_author_affiliation, corr_author_affiliation_country = None, None
            corr_author_has_email = corr_author.select('a[href^=mailto]')
            corr_author_email = corr_author_has_email[0].text if corr_author_has_email else None
            corr_author_has_orcid = corr_author.select('a[id^=contrib-orcid]')
            corr_author_orcid = corr_author_has_orcid[0]['href'] if corr_author_has_orcid else None
            corr_author_info = [corr_author_name, corr_author_affiliation, corr_author_affiliation_country,
                               corr_author_email, corr_author_orcid]
        else:
            corr_author_info = [None, None, None, None, None]
    except AttributeError:
        print(corr_author_email_card)
    num_authors = len(soup.find_all("div", {"class": "info-card-name"}))
    
    author = [num_authors] + first_author_info + corr_author_info
    return author

# data availability statement variables
def has_das(soup):
    """
    True if the publication has a data availability statement
    """
    das = soup.find("h2", {"data-section-title": "DATA AVAILABILITY STATEMENT"})
    return True if das is not None else False

def get_das_text(soup):
    """
    Get original text of data availability statement
    """
    if has_das(soup):
        das = soup.find("h2", {"data-section-title": "DATA AVAILABILITY STATEMENT"})
        das_text = soup.find("div", {"data-section-parent-id": das["id"]}).text
        das_text = das_text.strip()
        
    else:
        das_text = None
    return das_text

def get_das_repo(das):
    """
    Get online repository url if the data availability is online repository
    """
    if das is not None:
        das = das.strip()
        online = "All relevant data are available from an online repository or repositories"
        if das.startswith (online):
            repo = re.findall(r'\((https?://[^\s]+)\)', das.split(online)[1])
            if len(repo) == 0 :
                repo = re.findall(r'(https?://[^\s]+)', das.split(online)[1])
                if len(repo) == 0:
                    repo = re.findall(r'\(([^\s]+)\)', das.split(online)[1])
                    repo = repo[0] if len(repo) >0 else None
                else:
                    repo = repo[0]
            else:
                repo = repo[0]
        else:
            repo = None
    else:
        repo = None
    return repo

def get_das_info(soup):
    """
    Collect all relevant metadata about data availability statement
    """
    das_text = get_das_text(soup)
    das_type = None
    online = "All relevant data are available from an online repository or repositories"
    if das_text is not None and das_text.startswith(online):
        das_type = online
    else:
        das_type = das_text
    das_repo_url = get_das_repo(das_text)
    return [has_das(soup), das_text, das_type, das_repo_url]

# Keywords
def get_keywords(soup):
    """
    Get keywords of the publication
    """
    keywords = []
    has_keyword = soup.find("div", {"class": "kwd-group"})
    if has_keyword:
        for k in has_keyword.children:
            if not isinstance(k, bs4.element.NavigableString):
                keywords.append(k.text)
    return keywords

def get_paper_metadata(url):
    paper = requests.get(url, headers = {"User-Agent": "user"})
    soup = bs4.BeautifulSoup(paper.content, "html.parser")
    if "unusual traffic" in str(paper.content):
        print(url)
    metadata = get_supp(soup) + get_author_info(soup) + get_das_info(soup) + [get_keywords(soup)] 
    return metadata

def get_issues(start_vol, end_vol, end_issue):
    #end_issue = datetime.datetime.today().month - 1 ## last month from current date
    issues = []
    for vol in range(start_vol, end_vol+1):
        if vol == end_vol:
            if end_vol <= 10 and end_issue > 4:
                raise AssertionError("Volume 1-10 only has 4 issues each, end_issue has to be smaller or equal to 4")
            elif end_vol == 11 and end_issue > 6:
                raise AssertionError("Volume 11 only has 6 issues, end_issue has to be smaller or equal to 6") 
            else:
                assert raise AssertionError("End_issue has to be smaller or equal to 12") 
            for i in range(1, end_issue+1):
              issues.append(get_issue_page(root, vol, i))
        else:
            if vol <= 10:
                for i in range(1, 5):
                    issues.append(get_issue_page(root, vol, i))
            elif vol == 11:
                for i in range(1, 7):
                    issues.append(get_issue_page(root, vol, i))
            else:
                for i in range(1, 13):
                    issues.append(get_issue_page(root, vol, i))
    return issues

def build_overview_table(issues):
    article_tbl = []
    root = "https://iwaponline.com"
    journal_name = "Journal of Water, Sanitation & Hygiene for Development"
    for issue in issues:
        issue_page = issue["page"]
        soup = bs4.BeautifulSoup(issue_page, "html.parser")
        article_list = soup.find("div", {"id":"ArticleList"})
        articles = article_list.find_all("a", {"href":re.compile("/washdev/article/"), "class":None})
        for a in articles:
            url = root + a["href"]
            title = a.text
            data_resource_id = a.parent["data-resource-id-access"]
            published_year = issue["volume"]+2010
            article_tbl.append([data_resource_id, issue["volume"], issue["issue"], url, journal_name, title, published_year])
    overview_df = pd.DataFrame(article_tbl, columns=['paperid', 'volume', 'issue', 'url', 'journal','title', 'published_year'])
    return overview_df

def build_metadata_table(overview_df):
    metadata = []
    for url in tqdm(overview_df["url"]):
        metadata.append(get_paper_metadata(url))
    metadata_columns = ["is_supp", "num_supp", "supp_file_type", "supp_link",
                    "num_authors", "first_author_name", "first_author_affiliation", "first_author_affiliation_country", 
                    "first_author_email", "first_author_orcid", "correspondence_author_name", "correspondence_author_affiliation",
                    "correspondence_author_affiliation_country", "correspondence_author_email", "correspondence_author_orcid",
                    "has_das", "das", "das_type", "das_repo_url", "keywords"]
    metadata_df = pd.DataFrame(metadata, columns=metadata_columns)
    return metadata_df

if __name__ == "__main__":
    # Main scraping
    headers = {
        'User-Agent': 'user'
    } # need a header for authentification, otherwise get 403 back
    root = "https://iwaponline.com/washdev/issue" # Structure: https://iwaponline.com/washdev/issue/volumne number/issue number

    # volume 1 (2011) - volume 13 (2023)
    ## Get all issues, v0.0.1 access by volume 13 issue 11 
  
    issues = get_issues(start_vol = 1, end_vol = 13, end_issue = 11)  
    overview_df = build_overview_table(issues)
    metadata_df = build_metadata_table(overview_df)
  
    # Concatenate overview and metadata
    washdev_df = pd.concat([overview_df, metadata_df], axis=1)
    # washdev_df.to_csv("path/to/rawdata-directory/washdev.csv")
    # TODO: update from current data, so we don't need to scrape all issues\

File: data-raw/washdev_scraping.R
# Scrape article metadata from the Journal of Water, Sanitation and Hygiene
# for Development (https://iwaponline.com/washdev).
#
# R port of inst/python/washdev_scraping_selenium.py (issue #11). Uses a
# headless Chrome session via {chromote} because iwaponline.com sits behind
# Cloudflare and blocks plain HTTP clients. Overriding the "HeadlessChrome"
# user agent is enough to pass the challenge.
#
# The scraper is incremental: it reads data-raw/washdev.csv, determines the
# last volume and issue present, scrapes only newer issues, and appends.
# Progress is written back to the CSV after every issue, so an interrupted
# run can be resumed by sourcing the script again.
#
# Differences from the Python version, both deliberate:
# - A `doi` column is collected from the citation_doi meta tag (new).
# - When supplements have mixed file types, all types are recorded joined
#   by " & " (one per file, e.g. "docx & xlsx") instead of the literal
#   "misc", which previously required manual repair in data_processing.R.
#
# Run from the package root:
#   Rscript data-raw/washdev_scraping.R

library(chromote)
library(rvest)
library(xml2)
library(dplyr)
library(purrr)
library(stringr)
library(readr)
library(tibble)

JOURNAL_NAME <- "Journal of Water, Sanitation & Hygiene for Development"
SITE_ROOT <- "https://iwaponline.com"
ISSUE_ROOT <- "https://iwaponline.com/washdev/issue"
DAS_ONLINE_BOILERPLATE <-
  "All relevant data are available from an online repository or repositories"

# Browser session ---------------------------------------------------------

start_browser <- function() {
  session <- ChromoteSession$new()
  ua <- session$Browser$getVersion()$userAgent
  session$Network$setUserAgentOverride(
    userAgent = gsub("HeadlessChrome", "Chrome", ua)
  )
  session
}

#' Navigate to a URL and return the rendered HTML.
#'
#' Polls until the document is complete and no Cloudflare challenge text is
#' present, then sleeps 1-3 seconds (as the Python scraper did) to stay
#' polite. Returns NA_character_ if the page never settles.
fetch_page <- function(session, url, max_polls = 15L, poll_s = 3) {
  session$Page$navigate(url, wait_ = FALSE)
  html <- NA_character_
  for (i in seq_len(max_polls)) {
    Sys.sleep(poll_s)
    ready <- session$Runtime$evaluate("document.readyState")$result$value
    html <- session$Runtime$evaluate(
      "document.documentElement.outerHTML"
    )$result$value
    challenged <- grepl("Just a moment|Verifying you are human", html)
    if (identical(ready, "complete") && !challenged) break
    if (i == max_polls) {
      warning("Page did not settle: ", url, call. = FALSE)
      html <- NA_character_
    }
  }
  Sys.sleep(runif(1, 1, 3))
  html
}

# Small helpers ------------------------------------------------------------

node_exists <- function(node) !inherits(node, "xml_missing")

#' Format a character vector the way pandas wrote Python lists to the raw
#' CSV, e.g. "['a', 'b']" or "[]", so old and new rows share one format and
#' data_processing.R parses both identically.
py_list <- function(x) {
  x <- x[!is.na(x)]
  if (length(x) == 0) return("[]")
  paste0("['", paste(x, collapse = "', '"), "']")
}

#' First text node directly inside a node, stripped. Mirrors BeautifulSoup's
#' `.next` on an element whose first child is text.
first_text <- function(node) {
  txt <- xml_find_first(node, "./text()[normalize-space()]")
  if (inherits(txt, "xml_missing")) return(NA_character_)
  str_trim(xml_text(txt))
}

# Issue schedule -----------------------------------------------------------

#' Number of issues in a volume. Volumes 1-10 (2011-2020) had 4 issues,
#' volume 11 (2021) had 6, volume 12 (2022) onward is monthly.
issues_in_volume <- function(volume) {
  if (volume <= 10) 4L else if (volume == 11) 6L else 12L
}

# Issue table of contents --------------------------------------------------

#' Parse one issue's table-of-contents page into the overview columns:
#' one row per article with paperid, volume, issue, url, journal, title,
#' published_year. Returns a zero-row tibble when the page has no article
#' list (an issue that does not exist yet).
parse_issue_page <- function(html, volume, issue) {
  empty <- tibble(
    paperid = integer(), volume = integer(), issue = integer(),
    url = character(), journal = character(), title = character(),
    published_year = integer()
  )
  if (is.na(html)) return(empty)
  page <- read_html(html)
  article_list <- html_element(page, "#ArticleList")
  if (!node_exists(article_list)) return(empty)
  links <- html_elements(article_list, "a[href*='/washdev/article/']")
  # The Python scraper kept only <a> tags without a class attribute; the
  # classed ones are PDF/citation links pointing at the same articles
  links <- links[is.na(html_attr(links, "class"))]
  if (length(links) == 0) return(empty)
  tibble(
    paperid = map_chr(links, \(a) xml_attr(xml_parent(a), "data-resource-id-access")),
    volume = volume,
    issue = issue,
    url = paste0(SITE_ROOT, html_attr(links, "href")),
    journal = JOURNAL_NAME,
    title = str_trim(html_text(links)),
    published_year = volume + 2010L
  )
}

# Article metadata ---------------------------------------------------------

parse_supp <- function(page) {
  supp_nodes <- html_elements(page, "div.dataSuppLink")
  supp_types <- map_chr(supp_nodes, function(node) {
    # The file type is the text after the download link: "- docx file"
    txt <- str_trim(paste(xml_text(xml_find_all(node, "./text()")), collapse = " "))
    tokens <- str_split_1(str_trim(txt), "\\s+")
    if (length(tokens) >= 2) tokens[2] else NA_character_
  })
  supp_type <- if (length(supp_types) == 0) {
    NA_character_
  } else {
    paste(supp_types, collapse = " & ")
  }
  list(
    is_supp = node_exists(html_element(page, "h2#supplementary-data")),
    num_supp = length(supp_nodes),
    supp_file_type = supp_type,
    supp_url = py_list(map_chr(
      supp_nodes, \(node) html_attr(html_element(node, "a"), "href")
    ))
  )
}

#' Extract name, affiliation, country, email, and ORCID from one author
#' info card (a div.info-card-author node).
parse_author_card <- function(card) {
  name <- first_text(html_element(card, "div.info-card-name"))
  aff_node <- html_element(card, "div.aff")
  if (node_exists(aff_node)) {
    # The affiliation div may open with a <span> marker; the affiliation
    # text is then the node right after it
    span <- html_element(aff_node, "span")
    aff <- if (node_exists(span)) {
      xml_text(xml_find_first(aff_node, "./span[1]/following-sibling::node()[1]"))
    } else {
      html_text(aff_node)
    }
    aff <- str_trim(aff)
    country <- last(str_split_1(aff, ", "))
    country <- str_split_1(country, " Email")[1]
  } else {
    aff <- NA_character_
    country <- NA_character_
  }
  email_node <- html_element(card, "a[href^='mailto']")
  orcid_node <- html_element(card, "a[id^='contrib-orcid']")
  list(
    name = name,
    affiliation = aff,
    affiliation_country = country,
    email = if (node_exists(email_node)) html_text(email_node) else NA_character_,
    orcid = if (node_exists(orcid_node)) html_attr(orcid_node, "href") else NA_character_
  )
}

parse_authors <- function(page) {
  na_author <- list(
    name = NA_character_, affiliation = NA_character_,
    affiliation_country = NA_character_, email = NA_character_,
    orcid = NA_character_
  )
  cards <- html_elements(page, "div.info-card-author")
  first <- if (length(cards) > 0) parse_author_card(cards[[1]]) else na_author
  corresp_marker <- html_element(page, "div.info-author-correspondence")
  corresp <- if (node_exists(corresp_marker)) {
    parse_author_card(xml_parent(corresp_marker))
  } else {
    na_author
  }
  list(
    num_authors = length(html_elements(page, "div.info-card-name")),
    first_author_name = first$name,
    first_author_affiliation = first$affiliation,
    first_author_affiliation_country = first$affiliation_country,
    first_author_email = first$email,
    first_author_orcid = first$orcid,
    correspondence_author_name = corresp$name,
    correspondence_author_affiliation = corresp$affiliation,
    correspondence_author_affiliation_country = corresp$affiliation_country,
    correspondence_author_email = corresp$email,
    correspondence_author_orcid = corresp$orcid
  )
}

#' Repository URL from the DAS text, for statements that start with the
#' standard online-repository sentence. Three regexes in decreasing order
#' of strictness, as in the Python version: a URL inside parentheses, any
#' URL, anything inside parentheses.
extract_das_repo <- function(das_text) {
  if (is.na(das_text) || !startsWith(das_text, DAS_ONLINE_BOILERPLATE)) {
    return(NA_character_)
  }
  rest <- str_sub(das_text, nchar(DAS_ONLINE_BOILERPLATE) + 1)
  for (pattern in c(
    "\\((https?://[^\\s]+)\\)",
    "(https?://[^\\s]+)",
    "\\(([^\\s)]+)\\)"
  )) {
    hit <- str_match(rest, pattern)[, 2]
    if (!is.na(hit)) return(hit)
  }
  NA_character_
}

parse_das <- function(page) {
  heading <- html_element(
    page, "h2[data-section-title='DATA AVAILABILITY STATEMENT']"
  )
  has_das <- node_exists(heading)
  das_text <- NA_character_
  if (has_das) {
    section_id <- html_attr(heading, "id")
    body <- html_element(
      page, sprintf("div[data-section-parent-id='%s']", section_id)
    )
    if (node_exists(body)) das_text <- str_trim(html_text(body))
  }
  das_type <- if (!is.na(das_text) && startsWith(das_text, DAS_ONLINE_BOILERPLATE)) {
    DAS_ONLINE_BOILERPLATE
  } else {
    das_text
  }
  list(
    has_das = has_das,
    das = das_text,
    das_type = das_type,
    das_repo_url = extract_das_repo(das_text)
  )
}

parse_keywords <- function(page) {
  group <- html_element(page, "div.kwd-group")
  if (!node_exists(group)) return(py_list(character()))
  py_list(html_text(xml_children(group)))
}

parse_doi <- function(page) {
  meta <- html_element(page, "meta[name='citation_doi']")
  if (node_exists(meta)) html_attr(meta, "content") else NA_character_
}

#' All article-level metadata for one article page, as a one-row tibble.
parse_article_page <- function(html) {
  page <- read_html(html)
  as_tibble(c(
    parse_supp(page),
    parse_authors(page),
    parse_das(page),
    list(keywords = parse_keywords(page), doi = parse_doi(page))
  ))
}

#' Fetch and parse one article. On failure returns a one-row tibble of NAs
#' so the issue keeps its remaining articles.
scrape_article <- function(session, url) {
  tryCatch(
    parse_article_page(fetch_page(session, url)),
    error = function(e) {
      warning("Failed to parse ", url, ": ", conditionMessage(e), call. = FALSE)
      parse_article_page("<html></html>")
    }
  )
}

# Raw CSV round trip -------------------------------------------------------

read_washdev_raw <- function(path) {
  read_csv(path, show_col_types = FALSE, name_repair = "unique_quiet") |>
    rename(row_index = 1)
}

#' Write the raw table back in the format pandas produced: an unnamed
#' index as the first column.
write_washdev_raw <- function(data, path) {
  write_csv(data, path, na = "")
  lines <- read_lines(path)
  lines[1] <- sub("^row_index", "", lines[1])
  write_lines(lines, path)
}

# Main ---------------------------------------------------------------------

#' Scrape all issues newer than the ones in the raw CSV and append them.
#' `end_vol` defaults to the volume implied by the current year. For the
#' final volume, scraping stops at the first issue that has no article
#' list (not yet published).
update_washdev_raw <- function(raw_path = "data-raw/washdev.csv",
                               end_vol = as.integer(format(Sys.Date(), "%Y")) - 2010L) {
  existing <- read_washdev_raw(raw_path)
  last_vol <- max(existing$volume)
  last_issue <- max(existing$issue[existing$volume == last_vol])
  message("Raw data covers up to volume ", last_vol, " issue ", last_issue)

  session <- start_browser()
  on.exit(session$parent$close(), add = TRUE)

  for (vol in last_vol:end_vol) {
    first_issue <- if (vol == last_vol) last_issue + 1L else 1L
    for (iss in seq_len(issues_in_volume(vol))) {
      if (iss < first_issue) next
      message("Scraping volume ", vol, " issue ", iss, " ...")
      overview <- parse_issue_page(
        fetch_page(session, paste(ISSUE_ROOT, vol, iss, sep = "/")),
        volume = vol, issue = iss
      )
      if (nrow(overview) == 0) {
        message("  no articles found, stopping volume ", vol)
        break
      }
      metadata <- map(overview$url, \(u) scrape_article(session, u)) |>
        list_rbind()
      new_rows <- bind_cols(overview, metadata) |>
        mutate(
          paperid = as.integer(paperid),
          row_index = max(existing$row_index) + row_number()
        )
      existing <- bind_rows(existing, new_rows)
      write_washdev_raw(existing, raw_path)
      message("  added ", nrow(new_rows), " articles (total ", nrow(existing), ")")
    }
  }
  invisible(existing)
}

# Runs only when executed directly (Rscript data-raw/washdev_scraping.R),
# not when the file is source()d for its functions
if (sys.nframe() == 0 && !interactive()) {
  update_washdev_raw()
}
